lib/harmony/harmonytask/taskTypeHandler.go

```diff
@@ -0,0 +1,284 @@
+package harmonytask
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/filecoin-project/lotus/lib/harmony/harmonydb"
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var log = logging.Logger("harmonytask")
+
+type taskTypeHandler struct {
+	TaskInterface
+	TaskTypeDetails
+	TaskEngine *TaskEngine
+	Count      atomic.Int32 /// locked by TaskEngine's mutex
+
+}
+
+func (h *taskTypeHandler) AddTask(extra func(TaskID, *harmonydb.Tx) bool) {
+	var tID TaskID
+	did, err := h.TaskEngine.db.BeginTransaction(h.TaskEngine.ctx, func(tx *harmonydb.Tx) bool {
+		// create taskID (from DB)
+		_, err := tx.Exec(`INSERT INTO harmony_task (name, added_by, posted_time) 
+			VALUES ($1, $2, CURRENT_TIMESTAMP) `, h.Name, h.TaskEngine.ownerID)
+		if err != nil {
+			log.Error("Could not insert into harmonyTask", err)
+			return false
+		}
+		err = tx.QueryRow("SELECT id FROM harmony_task ORDER BY update_time DESC LIMIT 1").Scan(&tID)
+		if err != nil {
+			log.Error("Could not select ID: ", err)
+		}
+		return extra(tID, tx)
+	})
+	if err != nil {
+		log.Error(err)
+	}
+	if !did {
+		return
+	}
+
+	if !h.considerWork([]TaskID{tID}) {
+		h.TaskEngine.bump(h.Name) // We can't do it. How about someone else.
+	}
+}
+
+func (h *taskTypeHandler) considerWork(ids []TaskID) (workAccepted bool) {
```

Comment:
Could add logs in places calling this to have some way to see what mechanisms are actually triggering work scheduling in production systems.

Would be either logs next to calls to considerWork, or some `reason string` param here that would get logged here

Reply:
Thanks! Added line 107. I think that's what you're looking for.